models/ed/R/write.configs.ed.R

```diff
@@ -428,6 +316,152 @@ remove.config.ED2 <- function(main.outdir = settings$outdir, settings) {
   }
 }
 #==================================================================================================#
+#' @name write.config.xml.ED2
+#' @title Write ED2 config.xml file
+#' @details Refactored by Alexey Shiklomanov to allow use in PEcAn RTM module.
+#' @export
+#' @param settings PEcAn settings file. Settings required for this script are: model$revision, model$config.header, constants
+#' @param trait.values List of trait values with which to replace defaults
+#' @param defaults List of defaults to process. Default = settings$constants
+#' @return R XML object containing full ED2 XML file
+#' @author David LeBauer, Shawn Serbin, Carl Davidson, Alexey Shiklomanov
+write.config.xml.ED2 <- function(settings, trait.values, defaults=settings$constants){
+
+  ## Find history file
+  ## TODO this should come from the database
+  histfile <- paste("data/history.r", settings$model$revision, ".csv", sep='')
+  if (file.exists(system.file(histfile, package="PEcAn.ED2"))) {
+    #print(paste("--- Using ED2 History File: ","data/history.r", settings$model$revision, ".csv", sep=''))
+    edhistory <- read.csv2(system.file(histfile, package="PEcAn.ED2"), sep=";", 
+                           stringsAsFactors=FALSE, dec='.')
+  } else {
+    #print("--- Using Generic ED2 History File: data/history.csv")
+    edhistory <- read.csv2(system.file("data/history.csv",  package="PEcAn.ED2"), sep=";",
+                           stringsAsFactors=FALSE, dec='.')
+  }
+
+  edtraits <- names(edhistory)
+  data(pftmapping)
+
+  ## Get ED2 specific model settings and put into output config xml file
+  xml <- listToXml(settings$model$config.header, 'config')
+  
+  for(i in seq_along(trait.values)){
+    group <- names(trait.values)[i]
+    if(group == "env"){
+      
+      ## set defaults from config.header
+      
+      ##
+      
+    } else {
+      # Make this agnostic to the way PFT names are defined in `trait.values` -- either directly as list names or as object "name" within each sublist is fine
+      if(group == "pft"){
+        pft <- trait.values[[i]]$name
+      } else {
+        pft <- group
+      }
+      # TODO: Not sure if this is how this is supposed to work, but idea is to check for pft.number in settings first
+      pft.number <- settings$constants[[pft]]$num
+      if(is.null(pft.number)){
+          pft.number <- pftmapping$ED[which(pftmapping == pft)]
+      }
+      if(length(pft.number) == 0){
+          logger.error(pft, 'was not matched with a number in settings$constants or pftmapping data. Consult the PEcAn instructions on defining new PFTs.')
+          stop('Unable to set PFT number')
+      }
+        # TODO: Also modify web app to not default to 1 
+
+      ## Get default trait values from ED history
+      vals <- as.list(edhistory[edhistory$num == pft.number,])
+
+      ## Convert trait values to ED units
+      converted.trait.values <- convert.samples.ED(trait.values[[i]])
```

Comment:
I think convert.samples.ED should come after you replace the constants from the settings, because it can be the case that they might also need conversion